src/scripts.c

```diff
@@ -75,11 +72,16 @@ static void update_triggers(bool block)
 		string_or_die(&block_flag, "");
 	}
 
-	string_or_die(&restart_cmd, "/usr/bin/systemctl %s restart update-triggers.target", block_flag);
-	ret = system(restart_cmd);
+	if (strcmp("/", path_prefix) == 0) {
+		string_or_die(&cmd, "%s %s %s", POST_UPDATE, reexec_flag, block_flag);
+	} else {
+		string_or_die(&cmd, "%s/%s %s %s %s", path_prefix, POST_UPDATE, path_prefix, reexec_flag, block_flag);
```

Comment:
So does this mean the executable must accept the `--rexec` flag to run `daemon-reexec` and the `--no-block` block flag to pass to `systemctl`? It seems like if this is the case it should be pretty well documented. This is placing very specific restrictions on what is supposed to be a configurable script that can be added by whoever builds swupd.

Reply:
Well it doesn't have to accept these options but the result will be odd in the case of systemd not being reexec'd or slower. This is why the scripts probably need to be in swupd itself in order to properly handle only updating if a file in the mandb changed for instance.

The script itself can of course not take any arguments and do whatever it wants still though.

Maybe there is a way to remove the use of swupd depending on rexec or no block though.

Reply:
Well, we need the capability for the script to block in the event that swupd will be re-execing itself over format bumps since we don't want to start another swupd update until the post-update scripts are completed. Could we just always block on post-updates?

Reply:
That's exactly the kind of requirement that makes me extra sure these should live with swupd and let each distro configure their plugins with the rest of the project.

We can do whatever we want here though, I am fine with however it looks for now.

Reply:
I definitely like the plugin idea since these scripts have to do certain things specific to swupd's behavior. Just having a poorly-defined requirement for the user to provide their own pre- and post-update scripts at a certain path seems difficult for anyone picking up swupd to reason out on their own.

Reply:
Indeed, we should define how a plugin system for swupd would work and then implement it and scrap this change once that is done. I'm inclined to use this as a temporary resolution to our current problem however. Thoughts?

Reply:
Hm. Looking at how this is implemented though I'm beginning to see it as a step backward because of the specific requirements placed on the provided scripts. This feels more like removing functionality than solving a problem since it makes it more difficult for someone to take swupd as it exists and stand it up.

I like the decoupling *idea* but I don't like how this makes swupd more difficult to use by someone outside the Clear ecosystem. Am I missing something that makes this change necessary now instead of taking the time to implement a plugin architecture the right way?

Reply:
Maybe a better *now* solution would be to keep the hardcoded `systemctl` calls in place as a fallback if the configure options are not set.

Reply:
We can do that by changing  lines 259 and 265 of the configure.ac file.

Reply:
It is adding requirements since there now needs to be a script in place. If a team is already using swupd and have figured out their own way to manage without using systemd they probably are already patching swupd. If they are using systemd and have setup their triggers.target accordingly they'll have to do work for this change. I'm not sure how much of an impact that will be in practice.

It may be better to have this change be moved in as a patch to the Clear Linux package as it is important there in order to properly build and update docker images and also allows populating image state directories removing boot time performance penalties. That way the upstream doesn't introduce a breaking change only to move again once a plugin system is in place.

Reply:
It isn't a breaking change if you leave the systemd calls in place as a fallback. Then Clear doesn't have to carry a patch internally that may have to be updated every time a new swupd release is made until the plugin feature is introduced.